### ISSUE TITLE
Documentation and examples

### ISSUE THREAD

**Issue by andyferris** [FugroRoames/TypedTables.jl#23]
This package needs a user guide.

**Comment by andyferris** [comment]
There's some work-in-progress using Documenter.jl and github-pages now - you can find this with the "latest" documentation badge from the README.

**Comment by c42f** [comment]
Hey Andy,

I wrote the first two doc review parts to you in an email... but thought I'd continue here (for the syntax highlighting etc). I'm up to IO:

## Input and output

* Why are headings Tables.jl and CSV.jl in italics?

### Tables.jl

I'm not sure what information you're trying to convey in the Tables.jl section.
Are you saying that Table and FlexTable integrate with it? It's not clear how
this is related to IO until you get to the section on CSV.

Perhaps talking about the how TypedTables relates to the Tables API should go
into the Table types section?

### CSV.jl

Now here's where it gets really practical. I'd suggest renaming this section
something like "Reading delimited text files into a Table", and include an
example with the actual data which can be pasted into the REPL, perhaps.
For that you can put the string containing the delimited data inline:

```julia
raw_data = IOBuffer("""
name,age
Alice,25
Bob,42
Charlie,37
""")

csvfile = CSV.File(raw_data, delim=',')

table = Table(delimfile)
```

BTW the following is an interesting read
https://github.com/JuliaData/CSV.jl/issues/340

**Comment by c42f** [comment]
Thoughts on the next section

# Basic data manipulation

## Mapping rows of data

### Using map

Extracting a column - perhaps too simplistic given that there's another much
better way to do this using `t.name`? I think you could probably cut this
example.

How about this example of adding a new column?

```julia
julia> map(row -> (row..., is_old=row.age > 40), t)
Table with 3 columns and 3 rows:
     name     age  is_old
   ┌─────────────────────
 1 │ Alice    25   false
 2 │ Bob      42   true
 3 │ Charlie  37   false
```

### Generators

This is nice.

> Generators and comprehensions also support filtering data and combining multiple datasets, which cover in Finding Data and Joining Data.

* *covered*
* Link to the *Finding Data* and *Joining Data* sections


### Preselection

You're right, I couldn't see the point of getproperty(:name) out of context.


## Finding data

Example can be expressed as

```julia
t[t.age .> 40]
```

I guess this is a frustration I'm having with a lot of the examples in the
mapping and finding sections — it's good that they're simple, but on the other
hand they're kinda *unrealistically simple* in the sense that you woudln't
express the code that way in practice. IMO the examples should be just
complicated enough to show idiomatic use. Easy to say, I know.

**Comment by c42f** [comment]
Random (likely useless) thought bubble — having just written that AcceleratedArrays is great because it decouples performance from query semantics — could we think of the types in `Table` in a similar way; extra data to improve performance, but which might be missing? Thus somehow folding `FlexTable` and `Table` together?

**Comment by andyferris** [comment]
Haha - that's kind of interesting, actually. Until now I've been thinking of `FlexTable` as the slower `Table`. Should we define `decellerate(t::Table) = FlexTable(t)`? :smile:

**Comment by andyferris** [comment]
And thank you very much Chris for the valuable feedback! (I now have to find the time to make some fixes).

> one thing which strikes me here is that you're not really documenting TypedTables per se? But the Andyverse of data analysis? Which makes it kind of odd that the documentation is in TypedTables.jl.

Yes... well that is kind-of true. They were developed quite specifically to work together - a kind of "native" and "Julian" relational algebra interface. *SplitApplyCombine* deserves better documentation of it's own (and I'd like to port it to `Base`).

> Another thing which strikes me is that the grouping and joining sections seem quite polished. I especially enjoyed the grouping, which looks like it would address some of my frustrations having used the DataFrames `groupby`.

Thanks for the feedback! It's fair to say that *SplitApplyCombine* exists specifically because there is no `Base.group`, so yeah this is the bit which I definitely feel the most strongly about and have thought about the longest.

**Comment by c42f** [comment]
> Should we define decellerate(t::Table) = FlexTable(t)?

Not *quite* what I was thinking ;-)  More like trying to define

```julia
const FlexTable{N} = Table{Placeholder, N, NamedTuple{<:Any, <:Tuple{Vararg{AbstractArray{<:Any,N}}}}}
```

where `Placeholder` might be `Nothing` or `NamedTuple` undecorated with column names and types. Or something. And trying to see if that can lead anywhere productive.

**Comment by andyferris** [comment]
Yes, agreed. I already began a rewrite to create a much more focussed tutorial. Interestingly, this starkly highlighted a couple of the (known) missing features, so I'm looking into these as I go.

I'm not certain how to phrase the left-over design explanation without it being just a rant. Anyway; iterate, iterate, iterate...

**Comment by c42f** [comment]
Yeah the explanation of a design is hard to write and make useful. The abstract design arises from a bunch of concrete use cases and practical constraints... but writing those down without any organization leads to a pile. On the other hand, remove them and it feels like you're writing fluff without justification. Kind of like a rant, yes!

Maybe it would help to try to name the dimensions of the "use case space"? A given design satisfies the needs of a bunch of use cases, and so fills out some nontrivial volume in that space. At the boundaries of the volume are some particular extrema which the design only just satisfies... are these the use cases which matter and are worth discussing to keep things concrete?

On the other hand there's the design space and performance spaces, which (looking at the literature) seems to be [more standard concepts](http://sci-hub.tw/10.1115/1.4001528). But for software design the design space seems rather high dimensional, poorly defined and combinatoric rather than continuous. Probably like most real world design problems...

Oh, I'm sure some [category theory](https://pdfs.semanticscholar.org/140b/b309ca4b41df77ffb05146eb1c80da2593d4.pdf) will help us out. (Um. I have only the vaguest idea of what that paper is proposing.)

**Comment by c42f** [comment]
Oops, got the link wrong... [here's](http://sci-hub.tw/10.1109/SYSCON.2018.8369508) the paper which talks about using category theory for Formal Design.

**Comment by andyferris** [comment]
Thanks. Unfortunately, I haven't the time to look over something so... dense... at the moment ;)

**Comment by c42f** [comment]
I do wonder whether `_.name` will get into 1.1. The issue of binding tightness is really thorny. Reading back on the issue, I'm rather dissatisfied with tight binding and Stefan's counter proposal seems better but is quite complicated and lacks an implementation.

**Comment by andyferris** [comment]
I agree. It also seems thorny enough to sink it (or at least delay it signficantly).

**Comment by c42f** [comment]
Well, you totally nerd sniped me with the underscores business... Now there's [MagicUnderscores.jl](https://github.com/c42f/MagicUnderscores.jl). You're the first to see it :-P
